testgrid.tahoe-lafs.org/tahoe-lafs-testgrid.nix

```diff
@@ -0,0 +1,98 @@
+# Define a NixOS module that sets up the Tahoe-LAFS test grid.
+{ config, pkgs, ... }:
+let
+  # Use upstream packaging.  The NixOS 21.05 package is broken (though
+  # master should already have a fix for that).  However, maybe we want to
+  # run bleeding edge on this deployment anyway.
+  package = pkgs.callPackage ./tahoe-lafs.nix { };
+in {
+  # Configure Tahoe to run here.
+  services.tahoe = {
+
+    # Run two introducers so folks can play around with the multi-introducer
+    # support if they want.
+    introducers = {
+      # Just have them listen on different ports.
+      alpha = {
+        inherit package;
+        nickname = "alpha-introducer";
+        tub.port = 5000;
+      };
+      beta = {
+        inherit package;
+        nickname = "beta-introducer";
+        tub.port = 5001;
+      };
+    };
+
+    # Run three storage nodes.  They all share available storage space on this
+    # system but having more than one makes it more interesting to run a
+    # client.  On a more realistic deployment these would all be run
+    # separately from other to make their failure modes as independent as
+    # possible.
+    nodes =
+    let
+      # XXX NixOS module doesn't support multi-introducer configuration.
+      introducer = "pb://fodk4doc64febdoxke3a4ddfyanz7ajd@tcp:157.90.125.177:5000/el4fo3rm2h22cnilukmjqzyopdgqxrd2";
+    in {
+      alpha = {
+        inherit package;
+        nickname = "alpha-storage";
+        # XXX NixOS module requires we configure a web port even if we don't
+        # want one.
+        web.port = 2002;
```

Comment:
These take Twisted strings, so I think this should read `tcp:2002:interface=localhost` to ensure we're only listening locally, at least.

Reply:
Tahoe does, indeed, but the NixOS service module takes integers and builds the endpoint description strings itself. :/  We should try to get this improved in NixOS so we can do something better here.

Unfortunately the string it builds is just `tcp:2002` so we *are* listening on all interfaces - but NixOS defaults to enabling a no-inbound firewall and I've only punched holes for the services we want.  I'll also push a rev that explicitly enables the firewall rather than relying on it being enabled by default.